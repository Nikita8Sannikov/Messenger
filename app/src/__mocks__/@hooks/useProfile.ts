import { type IPhoto } from "@custom-types/models.types";

const mockNotificationSettingsService = {
	soundEnabled: false,
	messageSound: false,
	friendRequestSound: false,
	changeSoundNotifications: jest.fn(),
};

const mockPhotosService = {
	photos: [] as IPhoto[],
	count: 0,
	getAllPhotos: jest.fn(),
	onClickPhoto: jest.fn(),
	addPhotoFromAvatar: jest.fn(),
	addPhotos: jest.fn(),
	deletePhoto: jest.fn(),
};

const mockUserDetailsService = {
	updateDetails: jest.fn(),
	syncUserDetails: jest.fn(),
	getFriendsText: jest.fn(),
	getSubscribersText: jest.fn(),
	getPhotosText: jest.fn(),
	getAudiosText: jest.fn(),
	getVideosText: jest.fn(),
};

const mockUserService = {
	photosService: mockPhotosService,
	detailsService: mockUserDetailsService,
	settingsService: mockNotificationSettingsService,
	id: "",
	firstName: "",
	thirdName: "",
	phone: "",
	email: "",
	fullName: "",
	avatarUrl: "",
	avatarCreateDate: "",
	updateUser: jest.fn(),
	syncInfo: jest.fn(),
	changeAvatar: jest.fn(),
};

const mockProfileService = {
	isMe: false,
	userService: mockUserService,
	photosService: mockPhotosService,
	deleteAccount: jest.fn(),
	onSetAvatar: jest.fn(),
	onDeleteAvatar: jest.fn(),
	onClickAvatar: jest.fn(),
	getFriends: jest.fn(),
	editInfo: jest.fn(),
};

const useProfile = jest.fn().mockReturnValue(mockProfileService);

export default useProfile;
export { 
	mockProfileService, 
	mockUserService, 
	mockPhotosService, 
	mockUserDetailsService, 
	mockNotificationSettingsService, 
};