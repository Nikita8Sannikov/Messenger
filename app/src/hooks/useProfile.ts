import useMainClient from "@hooks/useMainClient";

// Возврат сущности "Профиль" конкретного пользователя по переданному идентификатору
export default function useProfile(userId?: string) {
	const mainClient = useMainClient();
	return mainClient.getProfile(userId);
}
