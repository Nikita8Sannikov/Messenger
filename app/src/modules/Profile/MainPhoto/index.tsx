import { useState } from "react";
import { useTranslation } from "react-i18next";
import Grid from "@mui/material/Grid";
import Paper from "@mui/material/Paper";

import useProfile from "@hooks/useProfile";
import ChangeAvatarComponent from "@components/ui/ChangeAvatar";
import PhotoComponent from "@components/ui/Photo";

import "./main-photo.scss";

export default function MainPhoto() {
    const [loading, setLoading] = useState(false);

    const { t } = useTranslation();
    const profile = useProfile();
    
    // Клик по своей аватарке
    const clickHandler = () => {
        profile.onClickAvatar();
    };

    // Установка/обновление своего аватара
    const changeHandler = (updateOptions: { id: string; newAvatarUrl: string; newPhotoUrl: string; }) => {
        profile.onSetAvatar(updateOptions);
    };

    // Удаление аватара
    const deleteHandler = () => {
        profile.onDeleteAvatar(setLoading);
    };

    return <Grid item>
        <Paper className="main-photo paper-block">
            <PhotoComponent
                src={profile.user.avatarUrl}
                alt="user-avatar"
                clickHandler={clickHandler}
                deleteHandler={deleteHandler}
            />

            <div className="main-photo__edit-button">
                <ChangeAvatarComponent
                    labelText={t("profile-module.change")}
                    loading={loading}
                    mustAuth
                    onChange={changeHandler}
                    setLoading={setLoading}
                />
            </div>
        </Paper>
    </Grid>
};