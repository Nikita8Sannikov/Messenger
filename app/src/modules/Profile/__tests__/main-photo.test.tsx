import { fireEvent, render, screen } from "@testing-library/react";

import MainPhoto from "@modules/profile/main-photo";
import useProfile from "../../../__mocks__/@hooks/useProfile";
import useProfileStore from "../../../__mocks__/@store/profile";

jest.mock("@service/i18n", () => ({
	t: (key: string) => key,
}));

describe("MainPhoto", () => {
	const mockProfile = {
		isMe: true,
		userService: {
			avatarUrl: "test-avatar.jpg",
		},
		onClickAvatar: jest.fn(),
		onSetAvatar: jest.fn(),
		onDeleteAvatar: jest.fn(),
	};

	beforeEach(() => {
		useProfileStore.getState().isDeleteAvatarLoading = false;
		useProfileStore.getState().setDeleteAvatarLoading = jest.fn();
		(useProfile as jest.Mock).mockReturnValue(mockProfile);
	});

	it("renders correctly", () => {
		const { container } = render(<MainPhoto />);
		expect(container).toMatchSnapshot();
	});

	it("renders avatar with correct src", () => {
		render(<MainPhoto />);
		const avatar = screen.getByTestId("image");
		expect(avatar).toHaveAttribute("src", "test-avatar.jpg");
	});

	it("shows change avatar button for own profile", () => {
		render(<MainPhoto />);
		const changeButton = screen.getByText("profile-module.change");
		expect(changeButton).toBeInTheDocument();
	});

	it("does not show change avatar button for other user's profile", () => {
		(useProfile as jest.Mock).mockReturnValue({
			...mockProfile,
			isMe: false,
		});

		render(<MainPhoto />);
		const changeButton = screen.queryByTestId("change-avatar-button");
		expect(changeButton).not.toBeInTheDocument();
	});

	it("calls onClickAvatar when avatar is clicked", () => {
		render(<MainPhoto />);
		const avatar = screen.getByTestId("image");
		fireEvent.click(avatar);
		expect(mockProfile.onClickAvatar).toHaveBeenCalled();
	});

	it("calls onDeleteAvatar when delete is triggered", () => {
		const mockOnDeleteAvatar = jest.fn();
		(useProfile as jest.Mock).mockReturnValue({
			...mockProfile,
			onDeleteAvatar: mockOnDeleteAvatar,
		});

		render(<MainPhoto />);
		const avatar = screen.getByTestId("image");
		fireEvent.contextMenu(avatar);
		expect(mockOnDeleteAvatar).not.toHaveBeenCalled();
	});
});
