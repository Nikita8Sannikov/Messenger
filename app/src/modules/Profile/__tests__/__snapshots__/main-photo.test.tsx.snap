// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPhoto renders correctly 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row main-photo css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 main-photo__container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="photo"
      >
        <img
          alt="user-avatar"
          class="photo__image"
          data-testid="image"
          loading="eager"
          src="test-avatar.jpg"
        />
      </div>
      <div
        class="main-photo__container__edit-button"
      >
        <label
          class="input-label"
          for="change-avatar-profile-module.change"
        >
          <input
            accept="image/*"
            data-testid="input-image"
            hidden=""
            id="change-avatar-profile-module.change"
            type="file"
          />
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary input-label__button css-wyro15-MuiButtonBase-root-MuiButton-root"
            id=":r0:"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-loadingWrapper"
              style="display: contents;"
            />
            <p
              class="MuiTypography-root MuiTypography-caption input-label__button__text css-mpaqru-MuiTypography-root"
            >
              profile-module.change
            </p>
          </button>
        </label>
      </div>
    </div>
  </div>
</div>
`;
